cql3/statements/alter_role_statement.hh

```diff
@@ -31,6 +31,7 @@ public:
     alter_role_statement(const cql3::role_name& name, const role_options& options)
                 : _role(name.to_string())
                 , _options(std::move(options)) {
+        needs_guard = true;
```

Comment:
Do we need to do it unconditionally? Shouldn't we only set it to true if we are in auth-on-raft mode?

Reply:
By other places I assume you mean the schema altering statement. It's kind of a special case because schema on raft is automatically enabled after the cluster creates group 0. Before creating group 0, the raft_group0_client interface can be still be used (i.e. `start_operation` and friends) but operates in a mode which does not use raft at all.

Here, you are imposing a requirement on non-auth-for-raft mode such that changes to auth now always require quorum of nodes to be alive (because of the need to start a group0 operation).

Reply:
That should still be fine because with default cassandra role queries are done with CL QUORUM anyway. Moreover I don't think anyone would see that it's a good idea to fiddle with auth when there is no quorum (except maybe disabling it in the config).

Reply:
Hmm, right. Considering that the legacy system_auth keyspace has simple replication strategy (AFAIK starting with RF=1, we recommend increasing to RF=3) then operations in legacy will actually start breaking much earlier before we lose clusterwide quorum.

I also see that and understand, again, plumbing a reference to the config might not be easy.

@gleb-cloudius what do you think? I'm fine with it if you are.

Reply:
Don't we have an easy way to subscribe to feature enablement event? How hard it is to pass the subscription to the statement code? OTOH isn't the "problem" only exists during cluster upgrade? Users cannot opt out from auth-raft node, right? If yes, then we can just ignore it IMO.

Reply:
the migration is not coded yet (so I guess there is no way in this patch to subscribe to anything), but the plan is to enable it automatically, if possible with an option to force rollback in case something goes wrong - but I am not sure if this is possible with current feature service (probably not?)

Reply:
We don't have a cluster feature for auth-on-raft (yet?), we only have an experimental feature in config.

In any case, the main problem seems to be that passing a reference to the feature (whatever the kind of the feature is) to all the relevant statements' constructors might be tedious.

> Users cannot opt out from auth-raft node, right? If yes, then we can just ignore it IMO.

I think the plan was to implement upgrade from legacy auth to auth-on-raft via the topology coordinator, and upgrade to topology coordinator is manually triggered. Does not running the upgrade count as "opt-out"?

Reply:
Ah, in this series the functionality is hidden behind experimental flag. A not enabled experimental feature should not have such big impact on the functionality. Looking at the code will it be possible to add the flag to `cql_config` which is a part of each `query_options`?

> if possible with an option to force rollback in case something goes wrong - but I am not sure if this is possible with current feature service (probably not?)

With features it is possible to rollback only before last node is upgraded and hence before a feature is enabled.

Reply:
It's possible to add flag to  cql_config but statement constructors doesn't have it available.

Reply:
Right. I do not know this area of the cql3 code, but it looks like statements are created deep in the parser code. May be we can do something similar to `cql_statement::raw_cql_statement` which is set in query_processor just after a statement is parsed.

Reply:
this is already after statement is created, so to distinguish you'd need to do dynamic_cast and then clear need_guard if auth-v2 is off and statement is certain type

Reply:
At the first version of the patches that introduced `needs_guard` it was a virtual function, not a flag. May be it is time to get back to it.

Reply:
Not sure how it helps?

Reply:
It can then return true of false depending on a config which you can attach after a statement is created.

Reply:
don't see where config would be directly attached to statement, but as I understand we can do it like this:
`statement->needs_guard(cql_config)` or perhaps better whole query_processor& qp as an argument? It anyway depends on it for `::execute(query_processor& qp,` calls

Reply:
> don't see where config would be directly attached to statement,
Look at how it is done with `raw_cql_statement`.

> but as I understand we can do it like this: `statement->needs_guard(cql_config)` or perhaps better whole query_processor& qp as an argument? It anyway depends on it for `::execute(query_processor& qp,` calls

Even better.